File: data/examples/alternative-classes-different-interfaces.cpp
#include <fstream>
#include <string>
#include <streambuf>

class FileReader {
public:
    std::string readFile(const std::string& filename) {
        std::ifstream file(filename);
        std::string content((std::istreambuf_iterator<char>(file)),
                           std::istreambuf_iterator<char>());
        return content;
    }
    
    bool fileExists(const std::string& filename) {
        std::ifstream file(filename);
        return file.good();
    }
    
    size_t getFileSize(const std::string& filename) {
        std::ifstream file(filename, std::ios::binary | std::ios::ate);
        return file.tellg();
    }
};

class DocumentLoader {
public:
    // 功能相同但方法名不同
    std::string loadDocument(const std::string& path) {
        std::ifstream file(path);
        std::string content((std::istreambuf_iterator<char>(file)),
                           std::istreambuf_iterator<char>());
        return content;
    }
    
    bool documentExists(const std::string& path) {
        std::ifstream file(path);
        return file.good();
    }
    
    // 参数顺序也不同
    long getDocumentSize(const std::string& path) {
        std::ifstream file(path, std::ios::binary | std::ios::ate);
        return file.tellg();
    }
};

File: data/examples/alternative-classes-different-interfaces.java
import java.io.*;
import java.nio.file.*;

class FileReader {
    public String readFile(String filename) throws IOException {
        return new String(Files.readAllBytes(Paths.get(filename)));
    }
    
    public boolean fileExists(String filename) {
        return Files.exists(Paths.get(filename));
    }
    
    public long getFileSize(String filename) throws IOException {
        return Files.size(Paths.get(filename));
    }
}

class DocumentLoader {
    // Same functionality but different method names
    public String loadDocument(String path) throws IOException {
        return new String(Files.readAllBytes(Paths.get(path)));
    }
    
    public boolean documentExists(String path) {
        return Files.exists(Paths.get(path));
    }
    
    // Different parameter order and return type
    public int getDocumentSize(String path) throws IOException {
        return (int) Files.size(Paths.get(path));
    }
}
